src/dtls13.c

```diff
@@ -0,0 +1,996 @@
+                                    word16 output_size,
+                                    word16 length,
+                                    enum HandShakeType handshake_type,
+                                    int hash_output)
+{
+    word16 record_header_length;
+    word16 record_length;
+    byte is_protected;
+    int send_length;
+    byte *msg;
+    int ret;
+
+
+    if (output_size < length)
+        return BUFFER_ERROR;
+
+    is_protected = dtls13_type_is_encrypted(handshake_type);
+    record_header_length = dtls13_get_rl_header_length(ssl, is_protected);
+
+    if (length <= record_header_length)
+            return BUFFER_ERROR;
+
+    record_length = length - record_header_length;
+
+    /* plain text (ClientHello/ServerHello/ACK) */
+    if (!is_protected) {
+
+        ret = dtls13_rl_add_plaintext_header(ssl,
+                                             output, handshake, record_length);
+        if (ret != 0)
+            return ret;
+
+    } else {
+        /* protected message  */
+        msg = output + record_header_length;
+
+        if (length <= record_header_length)
+            return BUFFER_ERROR;
+
+        /* FIXME: ignore async ok flag for now */
+        send_length = BuildTls13Message(ssl, output,
+                                        output_size,
+                                        msg, record_length, handshake,
+                                        hash_output, 0, 1);
+        if (send_length < 0)
+            return send_length;
+
+        length = send_length;
+    }
+
+    DtlsSEQIncrement(ssl, CUR_ORDER);
+
+    if (!frag_is_in_output_buffer(ssl, output))
+        return dtls13_send_frag_from_buffer(ssl, output, length);
+
+    ssl->buffers.outputBuffer.length += length;
+    return SendBuffered(ssl);
+}
+
+static void dtls13_free_fragments_buffer(WOLFSSL *ssl)
+{
+    XFREE(ssl->dtls13_fragments_buffer.buffer, ssl->heap,
+          DYNAMIC_TYPE_TEMP_BUFFER);
+    ssl->dtls13_fragments_buffer.buffer = NULL;
+    ssl->dtls13_sending_fragments = 0;
+    ssl->dtls13_message_length = ssl->dtls13_frag_offset = 0;
+}
+
+static int dtls13_send_fragmented_internal(WOLFSSL *ssl)
+{
+    int frag_length, rl_header_length;
+    int remaining_size, max_fragment;
+    int record_length;
+    byte is_encrypted;
+    byte *output;
+    int ret;
+
+    is_encrypted = dtls13_type_is_encrypted(ssl->dtls13_frag_handshake_type);
+    rl_header_length = dtls13_get_rl_header_length(ssl,is_encrypted);
+    max_fragment = wolfSSL_GetMaxFragSize(ssl, MAX_RECORD_SIZE);
+
+    remaining_size = ssl->dtls13_message_length - ssl->dtls13_frag_offset;
+
+    while(remaining_size > 0) {
+
+        frag_length = max_fragment - rl_header_length -
+            DTLS_HANDSHAKE_HEADER_SZ;
+
+        if (is_encrypted)
+            frag_length -= cipherExtraData(ssl);
+
+        if (frag_length > remaining_size)
+            frag_length = remaining_size;
+
+        ret = CheckAvailableSize(ssl, max_fragment);
```

Comment:
I think that `max_fragment` is equal to `rl_header_length + frag_length +  DTLS_HANDSHAKE_HEADER_SZ`,    because `frag_length` is calculated as: `frag_length = max_fragment -  rl_header_length - DTLS_HANDSHAKE_HEADER_SZ`. Of course if `max_fragment` takes in account cipher data, this must be changed to use all the available space.